File: 1046_LastStoneWeight.cpp
#include <local_leetcode.hpp>

class Solution {
  public:
    int lastStoneWeight(vector<int>& stones) {
      priority_queue<int> heap(stones.begin(), stones.end());

      while (heap.size() > 1) {
        int x = heap.top(); heap.pop();
        int y = heap.top(); heap.pop();

        if (x != y)
          heap.push(abs(x - y));
      }

      return heap.empty() ? 0 : heap.top();
    }
};


int main() {
  EXECS(Solution::lastStoneWeight);
  return 0;
}

File: src/1046_LastStoneWeight.java
import java.util.Arrays;
import java.util.PriorityQueue;

class LastStoneWeight {
  public int lastStoneWeight(int[] stones) {
    Arrays.sort(stones);
    int l=0, r=0;
    // unfinished
    return Math.abs(l-r);
  }

  public int lastStoneWeightHeap(int[] stones) {
    PriorityQueue<Integer> heap = new PriorityQueue<>((a, b) -> b-a);
    for (int w : stones)
      heap.offer(w);

    while (heap.size()>1) {
      int p = heap.poll();
      int q = heap.poll();
      if (p!=q)
        heap.offer(Math.abs(p-q));
    }
    return heap.isEmpty()?0:heap.poll();
  }
}